In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import nivapy3 as nivapy
import numpy as np
import pandas as pd
import seaborn as sn
import teotil2 as teo
import useful_rid_code as rid

sn.set_context("notebook")

# RID

## Estimating loads in unmonitored regions (parameterised)

This notebook is "parameterised" for use with Papermill. The cell below has the tag `parameters`, which means the entire notebook can be called from `01_recalculate_ospar_1990-2016_main.ipynb`.

The [TEOTIL2 model](https://nivanorge.github.io/teotil2/) is used to estimate loads in unmonitored areas. We know the regine ID for each of the 155 stations where water chemistry is measured, and we also know which OSPAR region each monitoring site drains to. We want to use observed data to estimate loads upstream of each monitoring point, and modelled data elsewhere.

In [2]:
# This cell is tagged 'parameters' for use with Papermill
# https://papermill.readthedocs.io/en/latest/index.html
year = 1990
user = ""
pw = ""

In [3]:
# Parameters
user = "jes"
pw = "BeakabusNov21.."
year = 2006


## 1. Generate model input file

In [4]:
# Parameters of interest
par_list = ["Tot-N", "Tot-P"]

# Path to TEOTIL2 "core" input data
teo_fold = r"../../../teotil2/data/core_input_data"

# Ouput path for model file
ann_input_csv = f"../../../teotil2/data/norway_annual_input_data/input_data_{year}.csv"

In [5]:
engine = rid.connect_to_nivabase(user=user, pw=pw)
df = teo.io.make_input_file(
    year, engine, teo_fold, ann_input_csv, mode="nutrients", par_list=par_list
)

Connection successful.


/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py:5176: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


## 2. Run model

In [6]:
%%time

# Run model
g = teo.model.run_model(ann_input_csv)

/opt/conda/lib/python3.9/site-packages/teotil2/model.py:67: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  df[f"trans_{par}"].between(0, 1, inclusive=True).all()


CPU times: user 6.14 s, sys: 62.1 ms, total: 6.21 s
Wall time: 6.26 s


## 3. Save results

In [7]:
# Save results as csv
res_csv = f"../../../teotil2/data/norway_annual_output_data/teotil2_results_{year}.csv"
df = teo.model.model_to_dataframe(g, out_path=res_csv)

df.head()

regine regine_ned  accum_agri_diff_tot-n_tonnes  \
0  001.1A2B   001.1A2A                      2.722851   
1  001.1A4D   001.1A4C                      0.000000   
2    001.1M     001.1L                      0.000000   
3  001.221Z   001.2210                      0.109854   
4  001.222Z   001.2220                      0.905250   

   accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
0                      0.181011                    0.033365   
1                      0.000000                    0.000000   
2                      0.000000                    0.000000   
3                      0.007303                    0.001346   
4                      0.060180                    0.011093   

   accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
0                    0.002762                      0.293833   
1                    0.000000                      0.000000   
2                    0.000000                      0.000000   
3                    0.000111                      0.011855   
4                    0.000918                      0.097689   

   accum_all_point_tot-p_tonnes  accum_all_sources_tot-n_tonnes  \
0                      0.036623                       15.084995   
1                      0.000000                        1.745975   
2                      0.000000                        4.057313   
3                      0.001478                        0.478399   
4                      0.012176                        2.378525   

   accum_all_sources_tot-p_tonnes  ...  local_ren_tot-n_tonnes  \
0                        0.557571  ...                     0.0   
1                        0.010679  ...                     0.0   
2                        0.036224  ...                     0.0   
3                        0.012651  ...                     0.0   
4                        0.088610  ...                     0.0   

   local_ren_tot-p_tonnes  local_runoff_mm/yr  local_spr_tot-n_tonnes  \
0                     0.0           409.50882                0.260468   
1                     0.0           477.76029                0.000000   
2                     0.0           477.76029                0.000000   
3                     0.0           409.50882                0.010509   
4                     0.0           409.50882                0.086596   

   local_spr_tot-p_tonnes  local_trans_tot-n  local_trans_tot-p  \
0                0.033861               1.00               1.00   
1                0.000000               0.81               0.26   
2                0.000000               0.86               0.31   
3                0.001366               1.00               1.00   
4                0.011258               1.00               1.00   

   local_urban_tot-n_tonnes  local_urban_tot-p_tonnes  local_vol_lake_m3  
0                   0.02205                   0.00315       1.788754e+08  
1                   0.00000                   0.00000       8.019497e+06  
2                   0.00000                   0.00000       3.866908e+07  
3                   0.00000                   0.00000       0.000000e+00  
4                   0.00000                   0.00000       0.000000e+00  

[5 rows x 54 columns]

In [8]:
# Save version with main catchments only
main_list = ["%03d." % i for i in range(1, 316)]
df2 = df.query("regine in @main_list")
df2.sort_values("regine", inplace=True)

# Save
main_csv = f"../../../Results/Unmon_loads/teotil2_results_{year}_main_catchs.csv"
df2.to_csv(main_csv, index=False, encoding="utf-8")

/opt/conda/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


## 4. Explore results

### 4.1. Total N and P

####  4.1.1. Identify areas with monitoring data

Where observations are available, we want to use them in preference to the model output. This means identifying all the catchments with observed data and substracting the model results for these locations. This is more complicated than it appears, because a small number of observed catchments are upstream of others, so subtracting all the loads for the 155 monitored catchments involves "double accounting", which we want to avoid. The first step is therefore to identify the downstream-most nodes for the monitored areas i.e. for the cases where one catchment is upstream of another, we just want the downstream node.

In [9]:
# Read station data
in_xlsx = r"../../../Data/RID_Sites_List_2017-2020.xlsx"
stn_df = pd.read_excel(in_xlsx, sheet_name="RID_All")
stn_df = stn_df.query("station_id != 38005")  # Ignore TROEMÅL2

# Get just cols of interest and drop duplicates
# (some sites are in the same regine)
stn_df = stn_df[["ospar_region", "nve_vassdrag_nr"]].drop_duplicates()

# Get catch IDs with obs data
obs_nds = set(stn_df["nve_vassdrag_nr"].values)

# Build network from input file
g, nd_list = teo.calib.build_calib_network(ann_input_csv, obs_nds)

# Get list of downstream nodes
ds_nds = []
for nd in g:
    # If no downstream nodes
    if g.out_degree(nd) == 0:
        # Node is of interest
        ds_nds.append(nd)

# Get just the downstream catchments
stn_df = stn_df[stn_df["nve_vassdrag_nr"].isin(ds_nds)]

#### 4.1.2. Sum model results for monitored locations

In [10]:
# Read model output
teo_df = pd.read_csv(res_csv)

# Join accumulated outputs to stns of interest
mon_df = pd.merge(
    stn_df, teo_df, how="left", left_on="nve_vassdrag_nr", right_on="regine"
)

# Groupby OSPAR region
mon_df = mon_df.groupby("ospar_region").sum()

# Get just accum cols
cols = [i for i in mon_df.columns if i.split("_")[0] == "accum"]
mon_df = mon_df[cols]

mon_df.head()

accum_agri_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    146.860444   
NORTH SEA                             2900.200796   
NORWEGIAN SEA2                        2941.061390   
SKAGERAK                             11073.282642   

                     accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                      4.980509                    2.540371   
NORTH SEA                               64.779256                   34.943085   
NORWEGIAN SEA2                          76.398308                   40.709881   
SKAGERAK                               215.873551                  102.034032   

                     accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                    0.213926                     82.064755   
NORTH SEA                              2.173150                    334.836099   
NORWEGIAN SEA2                         3.053550                    478.089315   
SKAGERAK                               5.504629                   3279.687383   

                     accum_all_point_tot-p_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                      7.648365   
NORTH SEA                               50.201968   
NORWEGIAN SEA2                          38.479649   
SKAGERAK                               114.213808   

                     accum_all_sources_tot-n_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                     3409.147435   
NORTH SEA                              11043.932615   
NORWEGIAN SEA2                         11151.616939   
SKAGERAK                               28545.077093   

                     accum_all_sources_tot-p_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                       53.644648   
NORTH SEA                                182.381327   
NORWEGIAN SEA2                           225.080764   
SKAGERAK                                 473.475836   

                     accum_anth_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    146.860444   
NORTH SEA                             2906.173366   
NORWEGIAN SEA2                        2957.528368   
SKAGERAK                             11235.217881   

                     accum_anth_diff_tot-p_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                      4.980509  ...   
NORTH SEA                               65.322733  ...   
NORWEGIAN SEA2                          78.527459  ...   
SKAGERAK                               233.351314  ...   

                     accum_nat_diff_tot-n_tonnes  accum_nat_diff_tot-p_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                  3180.222236                    41.015775   
NORTH SEA                            7802.923150                    66.856626   
NORWEGIAN SEA2                       7715.999256                   108.073656   
SKAGERAK                            14030.171829                   125.910714   

                     accum_q_m3/s  accum_ren_tot-n_tonnes  \
ospar_region                                                
LOFOTEN-BARENTS SEA    724.437505               24.441761   
NORTH SEA             1057.289958              157.705043   
NORWEGIAN SEA2        1421.853281              254.703811   
SKAGERAK              1753.662585             2142.609770   

                     accum_ren_tot-p_tonnes  accum_spr_tot-n_tonnes  \
ospar_region                                                          
LOFOTEN-BARENTS SEA                1.949012               55.082623   
NORTH SEA   

This table gives the **modelled** inputs to each OSPAR region from catchments for which we have observed data. We want to subtract these values from the overall modelled inputs to each region and substitute the observed data instead.

The trickiest part of this is that the OSPAR regions in the TEOTIL catchment network (files named `regine_{year}.csv`) don't exactly match the relevant OSPAR definitions for this analysis. This is because the "OSPAR boundaries" in the model include catchments draining to Sweden (as part of TEOTIL2 Metals - see [here](https://nivanorge.github.io/teotil2/pages/07_1000_lakes.html)), so instead of using them directly we need to aggregate based on vassdragsnummers.

#### 4.1.3. Group model output according to "new" OSPAR regions

In [11]:
# Define "new" OSPAR regions (ranges are inclusive)
os_dict = {
    "SKAGERAK": (1, 23),
    "NORTH SEA": (24, 90),
    "NORWEGIAN SEA2": (91, 170),
    "LOFOTEN-BARENTS SEA": (171, 247),
}

# Container for results
df_list = []

# Loop over model output
for reg in os_dict.keys():
    min_id, max_id = os_dict[reg]

    regs = ["%03d." % i for i in range(min_id, max_id + 1)]

    # Get data for this region
    df2 = teo_df[teo_df["regine"].isin(regs)]

    # Get just accum cols
    cols = [i for i in df2.columns if i.split("_")[0] == "accum"]
    df2 = df2[cols]

    # Add region
    df2["ospar_region"] = reg

    # Add to output
    df_list.append(df2)

# Build df
os_df = pd.concat(df_list, axis=0)

# Aggregate
os_df = os_df.groupby("ospar_region").sum()

os_df.head()

accum_agri_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    616.827188   
NORTH SEA                             6940.589810   
NORWEGIAN SEA2                        7963.301656   
SKAGERAK                             12712.198441   

                     accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                     25.503755                   10.569251   
NORTH SEA                              177.389053                   80.515432   
NORWEGIAN SEA2                         244.892879                  101.652087   
SKAGERAK                               278.083942                  112.948837   

                     accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                    0.866142                   7818.680733   
NORTH SEA                              5.832951                  14830.135960   
NORWEGIAN SEA2                         8.012104                  19208.394729   
SKAGERAK                               6.321836                   8994.146888   

                     accum_all_point_tot-p_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                   1574.900894   
NORTH SEA                             2759.317043   
NORWEGIAN SEA2                        3741.186208   
SKAGERAK                               301.308139   

                     accum_all_sources_tot-n_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                    15985.692902   
NORTH SEA                              37988.125245   
NORWEGIAN SEA2                         41721.877612   
SKAGERAK                               37134.328552   

                     accum_all_sources_tot-p_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                     1718.539859   
NORTH SEA                               3098.467641   
NORWEGIAN SEA2                          4202.268876   
SKAGERAK                                 749.909587   

                     accum_anth_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    626.799497   
NORTH SEA                             6980.407292   
NORWEGIAN SEA2                        7995.460446   
SKAGERAK                             12985.932910   

                     accum_anth_diff_tot-p_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                     26.907596  ...   
NORTH SEA                              182.457723  ...   
NORWEGIAN SEA2                         249.133322  ...   
SKAGERAK                               310.815909  ...   

                     accum_nat_diff_tot-n_tonnes  accum_nat_diff_tot-p_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                  7540.212672                   116.731369   
NORTH SEA                           16177.581993                   156.692876   
NORWEGIAN SEA2                      14518.022437                   211.949346   
SKAGERAK                            15154.248754                   137.785539   

                     accum_q_m3/s  accum_ren_tot-n_tonnes  \
ospar_region                                                
LOFOTEN-BARENTS SEA   1886.483721              869.729215   
NORTH SEA             2300.876526             2819.813498   
NORWEGIAN SEA2        2896.283239             2224.048993   
SKAGERAK              1871.942779             6467.735239   

                     accum_ren_tot-p_tonnes  accum_spr_tot-n_tonnes  \
ospar_region                                                          
LOFOTEN-BARENTS SEA              135.693531              318.701274   
NORTH SEA   

We can now calculate the unmonitored component by simply subtracting the values modelled upstream of monitoring stations from the overall modelled inputs to each OSPAR region.

#### 4.1.4. Estimate loads in unmonitored areas

In [12]:
# Calc unmonitored loads
unmon_df = os_df - mon_df

# Write output
out_csv = f"../../../Results/Unmon_loads/teotil2_raw_unmonitored_loads_{year}.csv"
unmon_df.to_csv(out_csv, encoding="utf-8", index_label="ospar_region")

unmon_df.round(0).astype(int).T

ospar_region                    LOFOTEN-BARENTS SEA  NORTH SEA  \
accum_agri_diff_tot-n_tonnes                    470       4040   
accum_agri_diff_tot-p_tonnes                     21        113   
accum_agri_pt_tot-n_tonnes                        8         46   
accum_agri_pt_tot-p_tonnes                        1          4   
accum_all_point_tot-n_tonnes                   7737      14495   
accum_all_point_tot-p_tonnes                   1567       2709   
accum_all_sources_tot-n_tonnes                12577      26944   
accum_all_sources_tot-p_tonnes                 1665       2916   
accum_anth_diff_tot-n_tonnes                    480       4074   
accum_anth_diff_tot-p_tonnes                     22        117   
accum_aqu_tot-n_tonnes                         6620      10661   
accum_aqu_tot-p_tonnes                         1399       2253   
accum_ind_tot-n_tonnes                            0        457   
accum_ind_tot-p_tonnes                            0         67   
accum_nat_diff_tot-n_tonnes                    4360       8375   
accum_nat_diff_tot-p_tonnes                      76         90   
accum_q_m3/s                                   1162       1244   
accum_ren_tot-n_tonnes                          845       2662   
accum_ren_tot-p_tonnes                          134        306   
accum_spr_tot-n_tonnes                          264        670   
accum_spr_tot-p_tonnes                           34         79   
accum_upstr_area_km2                          74535      35961   
accum_urban_tot-n_tonnes                         10         34   
accum_urban_tot-p_tonnes                          1          5   

ospar_region                    NORWEGIAN SEA2  SKAGERAK  
accum_agri_diff_tot-n_tonnes              5022      1639  
accum_agri_diff_tot-p_tonnes               168        62  
accum_agri_pt_tot-n_tonnes                  61        11  
accum_agri_pt_tot-p_tonnes                   5         1  
accum_all_point_tot-n_tonnes             18730      5714  
accum_all_point_tot-p_tonnes              3703       187  
accum_all_sources_tot-n_tonnes           30570      8589  
accum_all_sources_tot-p_tonnes            3977       276  
accum_anth_diff_tot-n_tonnes              5038      1751  
accum_anth_diff_tot-p_tonnes               171        77  
accum_aqu_tot-n_tonnes                   15402        31  
accum_aqu_tot-p_tonnes                    3255         7  
accum_ind_tot-n_tonnes                     770      1143  
accum_ind_tot-p_tonnes                     100        72  
accum_nat_diff_tot-n_tonnes               6802      1124  
accum_nat_diff_tot-p_tonnes                104        12  
accum_q_m3/s                              1474       118  
accum_ren_tot-n_tonnes                    1969      4325  
accum_ren_tot-p_tonnes                     276        86  
accum_spr_tot-n_tonnes                     528       204  
accum_spr_tot-p_tonnes                      67        22  
accum_upstr_area_km2                     68037      8629  
accum_urban_tot-n_tonnes                    16       112  
accum_urban_tot-p_tonnes                     2        15

#### 4.1.5. Aggregate values to required quantities

In [13]:
# Aggregate to match report
unmon_df["flow"] = unmon_df["accum_q_m3/s"] * 60 * 60 * 24 / 1000.0  # 1000s m3/day

unmon_df["sew_n"] = (
    unmon_df["accum_ren_tot-n_tonnes"] + unmon_df["accum_spr_tot-n_tonnes"]
)
unmon_df["sew_p"] = (
    unmon_df["accum_ren_tot-p_tonnes"] + unmon_df["accum_spr_tot-p_tonnes"]
)

unmon_df["ind_n"] = unmon_df["accum_ind_tot-n_tonnes"]
unmon_df["ind_p"] = unmon_df["accum_ind_tot-p_tonnes"]

unmon_df["fish_n"] = unmon_df["accum_aqu_tot-n_tonnes"]
unmon_df["fish_p"] = unmon_df["accum_aqu_tot-p_tonnes"]

unmon_df["diff_n"] = (
    unmon_df["accum_anth_diff_tot-n_tonnes"] + unmon_df["accum_nat_diff_tot-n_tonnes"]
)
unmon_df["diff_p"] = (
    unmon_df["accum_anth_diff_tot-p_tonnes"] + unmon_df["accum_nat_diff_tot-p_tonnes"]
)

new_df = unmon_df[
    ["flow", "sew_n", "sew_p", "ind_n", "ind_p", "fish_n", "fish_p", "diff_n", "diff_p"]
]

# Total for Norway
new_df.loc["NORWAY"] = new_df.sum(axis=0)

# Reorder rows
new_df = new_df.reindex(
    ["NORWAY", "LOFOTEN-BARENTS SEA", "NORTH SEA", "NORWEGIAN SEA2", "SKAGERAK"]
)

new_df.round().astype(int)

/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:723: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value, self.name)


flow  sew_n  sew_p  ind_n  ind_p  fish_n  fish_p  \
ospar_region                                                              
NORWAY               345457  11468   1003   2370    238   32714    6914   
LOFOTEN-BARENTS SEA  100401   1109    167      0      0    6620    1399   
NORTH SEA            107446   3332    385    457     67   10661    2253   
NORWEGIAN SEA2       127391   2498    343    770    100   15402    3255   
SKAGERAK              10219   4530    108   1143     72      31       7   

                     diff_n  diff_p  
ospar_region                         
NORWAY                32004     668  
LOFOTEN-BARENTS SEA    4840      98  
NORTH SEA             12449     207  
NORWEGIAN SEA2        11840     274  
SKAGERAK               2875      89

## 5. Other N and P species

Tore's procedure `RESA2.FIXTEOTILPN` defines simple correction factors for estimating PO4, NO3 and NH4 from total P and N. The table below lists the factors used.

|   Source    | Phosphate | Nitrate | Ammonium |
|:-----------:|:---------:|:-------:|:--------:|
|    Sewage   |     0.600 |   0.050 |    0.750 |
|   Industry  |     0.600 |   0.050 |    0.750 |
| Aquaculture |     0.690 |   0.110 |    0.800 |
|   Diffuse   |     0.246 |   0.625 |    0.055 |


In [14]:
# Dict of conversion factors
con_dict = {
    ("sew", "po4"): ("p", 0.6),
    ("ind", "po4"): ("p", 0.6),
    ("fish", "po4"): ("p", 0.69),
    ("diff", "po4"): ("p", 0.246),
    ("sew", "no3"): ("n", 0.05),
    ("ind", "no3"): ("n", 0.05),
    ("fish", "no3"): ("n", 0.11),
    ("diff", "no3"): ("n", 0.625),
    ("sew", "nh4"): ("n", 0.75),
    ("ind", "nh4"): ("n", 0.75),
    ("fish", "nh4"): ("n", 0.8),
    ("diff", "nh4"): ("n", 0.055),
}

# Apply factors
for src in ["sew", "ind", "fish", "diff"]:
    for spc in ["po4", "no3", "nh4"]:
        el, fac = con_dict[(src, spc)]
        new_df[src + "_" + spc] = fac * new_df[src + "_" + el]

new_df.round().astype(int).T

ospar_region  NORWAY  LOFOTEN-BARENTS SEA  NORTH SEA  NORWEGIAN SEA2  SKAGERAK
flow          345457               100401     107446          127391     10219
sew_n          11468                 1109       3332            2498      4530
sew_p           1003                  167        385             343       108
ind_n           2370                    0        457             770      1143
ind_p            238                    0         67             100        72
fish_n         32714                 6620      10661           15402        31
fish_p          6914                 1399       2253            3255         7
diff_n         32004                 4840      12449           11840      2875
diff_p           668                   98        207             274        89
sew_po4          602                  100        231             206        65
sew_no3          573                   55        167             125       226
sew_nh4         8601                  832       2499            1873      3397
ind_po4          143                    0         40              60        43
ind_no3          118                    0         23              38        57
ind_nh4         1777                    0        343             577       857
fish_po4        4771                  965       1555            2246         5
fish_no3        3599                  728       1173            1694         3
fish_nh4       26171                 5296       8529           12321        25
diff_po4         164                   24         51              68        22
diff_no3       20002                 3025       7781            7400      1797
diff_nh4        1760                  266        685             651       158

## 6. Other quantities

The model currently only considers N and P, but the project focuses on a wider range of parameters. For now, we simply assume that all measured inputs (`renseanlegg`, `industri` and `akvakultur`) for regines outside of catchments with measured data make it to the sea.

We only want data for catchments that are not monitored i.e. for regine IDs **not** in the graph created above.

In [15]:
engine = rid.connect_to_nivabase(user=user, pw=pw)

# The sql below uses a horrible (and slow!) hack to get around Oracle's
# 1000 item limit on IN clauses. See here for details:
# https://stackoverflow.com/a/9084247/505698
nd_list_hack = [(1, i) for i in nd_list]

sql = (
    "SELECT SUBSTR(a.regine, 1, 3) AS vassdrag, "
    "  a.type, "
    "  b.name, "
    "  b.unit, "
    "  SUM(c.value * d.factor) as value "
    "FROM RESA2.RID_PUNKTKILDER a, "
    "RESA2.RID_PUNKTKILDER_OUTPAR_DEF b, "
    "RESA2.RID_PUNKTKILDER_INPAR_VALUES c, "
    "RESA2.RID_PUNKTKILDER_INP_OUTP d "
    "WHERE a.anlegg_nr = c.anlegg_nr "
    "AND (1, a.regine) NOT IN %s "
    "AND d.in_pid = c.inp_par_id "
    "AND d.out_pid = b.out_pid "
    "AND c.year = %s "
    "GROUP BY SUBSTR(a.regine, 1, 3), a.type, b.name, b.unit "
    "ORDER BY SUBSTR(a.regine, 1, 3), a.type" % (tuple(nd_list_hack), year)
)

df = pd.read_sql(sql, engine)

# Tidy
df["par"] = df["type"] + "_" + df["name"] + "_" + df["unit"]
del df["name"], df["unit"], df["type"]

# Pivot
df = df.pivot(index="vassdrag", columns="par", values="value")
df.reset_index(inplace=True)

Connection successful.


In [16]:
def f(x):
    try:
        a = int(x)
        return a
    except:
        return -999


# Convert vassdrag to numbers
df["vass"] = df["vassdrag"].apply(f)

# Get just the main catchments
df = df.query("vass != -999")

df.head()

par vassdrag  INDUSTRI_As_tonn  INDUSTRI_Cd_tonn  INDUSTRI_Cr_tonn  \
0        001               NaN               NaN               NaN   
1        002          0.000153      2.832000e-05            0.1893   
2        003               NaN      5.000000e-07               NaN   
3        004               NaN               NaN               NaN   
4        005               NaN               NaN               NaN   

par  INDUSTRI_Cu_tonn  INDUSTRI_Hg_tonn  INDUSTRI_NH3_tonn  \
0                 NaN               NaN                NaN   
1            5.673222      7.401130e-03                NaN   
2                 NaN      1.300000e-07                NaN   
3                 NaN               NaN                NaN   
4                 NaN               NaN                NaN   

par  INDUSTRI_NH4-N_tonn  INDUSTRI_Ni_tonn  INDUSTRI_PCB_tonn  ...  \
0                    NaN               NaN                NaN  ...   
1                    NaN          0.931028                NaN  ...   
2                    NaN               NaN                NaN  ...   
3                    NaN               NaN                NaN  ...   
4                    NaN               NaN                NaN  ...   

par  RENSEANLEGG_Ni_tonn  RENSEANLEGG_PAH_tonn  RENSEANLEGG_PCB_tonn  \
0                    NaN                   NaN                   NaN   
1                0.13648               0.00682               0.00096   
2                0.04760                   NaN                   NaN   
3                    NaN                   NaN                   NaN   
4                    NaN                   NaN                   NaN   

par  RENSEANLEGG_Pb_tonn  RENSEANLEGG_S.P.M._tonn  RENSEANLEGG_TOC_tonn  \
0                0.00810                      NaN               160.856   
1                0.03839                  733.754               676.270   
2                0.00830                      NaN               235.118   
3                    NaN                      NaN                   NaN   
4                    NaN                      NaN                   NaN   

par  RENSEANLEGG_Tot-N_tonn  RENSEANLEGG_Tot-P_tonn  RENSEANLEGG_Zn_tonn  vass  
0                 104.29685                 2.03550               0.1664     1  
1                 477.98182                10.06891               4.0444     2  
2                 246.41176                 2.91400               0.2008     3  
3                  81.46572                 1.11121                  NaN     4  
4                  34.02384                 0.15901                  NaN     5  

[5 rows x 31 columns]

In [17]:
def f2(x):
    if x in range(1, 24):
        return "SKAGERAK"
    elif x in range(24, 91):
        return "NORTH SEA"
    elif x in range(91, 171):
        return "NORWEGIAN SEA2"
    elif x in range(171, 248):
        return "LOFOTEN-BARENTS SEA"
    else:
        return np.nan


# Assign main catchments to OSPAR regions
df["osp_reg"] = df["vass"].apply(f2)

df.head()

par vassdrag  INDUSTRI_As_tonn  INDUSTRI_Cd_tonn  INDUSTRI_Cr_tonn  \
0        001               NaN               NaN               NaN   
1        002          0.000153      2.832000e-05            0.1893   
2        003               NaN      5.000000e-07               NaN   
3        004               NaN               NaN               NaN   
4        005               NaN               NaN               NaN   

par  INDUSTRI_Cu_tonn  INDUSTRI_Hg_tonn  INDUSTRI_NH3_tonn  \
0                 NaN               NaN                NaN   
1            5.673222      7.401130e-03                NaN   
2                 NaN      1.300000e-07                NaN   
3                 NaN               NaN                NaN   
4                 NaN               NaN                NaN   

par  INDUSTRI_NH4-N_tonn  INDUSTRI_Ni_tonn  INDUSTRI_PCB_tonn  ...  \
0                    NaN               NaN                NaN  ...   
1                    NaN          0.931028                NaN  ...   
2                    NaN               NaN                NaN  ...   
3                    NaN               NaN                NaN  ...   
4                    NaN               NaN                NaN  ...   

par  RENSEANLEGG_PAH_tonn  RENSEANLEGG_PCB_tonn  RENSEANLEGG_Pb_tonn  \
0                     NaN                   NaN              0.00810   
1                 0.00682               0.00096              0.03839   
2                     NaN                   NaN              0.00830   
3                     NaN                   NaN                  NaN   
4                     NaN                   NaN                  NaN   

par  RENSEANLEGG_S.P.M._tonn  RENSEANLEGG_TOC_tonn  RENSEANLEGG_Tot-N_tonn  \
0                        NaN               160.856               104.29685   
1                    733.754               676.270               477.98182   
2                        NaN               235.118               246.41176   
3                        NaN                   NaN                81.46572   
4                        NaN                   NaN                34.02384   

par  RENSEANLEGG_Tot-P_tonn  RENSEANLEGG_Zn_tonn  vass   osp_reg  
0                   2.03550               0.1664     1  SKAGERAK  
1                  10.06891               4.0444     2  SKAGERAK  
2                   2.91400               0.2008     3  SKAGERAK  
3                   1.11121                  NaN     4  SKAGERAK  
4                   0.15901                  NaN     5  SKAGERAK  

[5 rows x 32 columns]

In [18]:
# Group by OSPAR region
df.fillna(0, inplace=True)
df = df.groupby("osp_reg").sum()
if 0 in df.index:
    df.drop(0, inplace=True)

# Total for Norway
df.loc["NORWAY"] = df.sum(axis=0)

# Join to model results
df = new_df.join(df)

# Get cols of interest
umod_cols = ["S.P.M.", "TOC", "As", "Pb", "Cd", "Cu", "Zn", "Ni", "Cr", "Hg"]
umod_cols = [
    "%s_%s_tonn" % (i, j) for i in ["INDUSTRI", "RENSEANLEGG"] for j in umod_cols
]
cols = list(new_df.columns) + umod_cols
cols.remove("RENSEANLEGG_TOC_tonn")
cols = [i for i in cols if i in df.columns]
df = df[cols]

df.round(0).astype(int).T

ospar_region             NORWAY  LOFOTEN-BARENTS SEA  NORTH SEA  \
flow                     345457               100401     107446   
sew_n                     11468                 1109       3332   
sew_p                      1003                  167        385   
ind_n                      2370                    0        457   
ind_p                       238                    0         67   
fish_n                    32714                 6620      10661   
fish_p                     6914                 1399       2253   
diff_n                    32004                 4840      12449   
diff_p                      668                   98        207   
sew_po4                     602                  100        231   
sew_no3                     573                   55        167   
sew_nh4                    8601                  832       2499   
ind_po4                     143                    0         40   
ind_no3                     118                    0         23   
ind_nh4                    1777                    0        343   
fish_po4                   4771                  965       1555   
fish_no3                   3599                  728       1173   
fish_nh4                  26171                 5296       8529   
diff_po4                    164                   24         51   
diff_no3                  20002                 3025       7781   
diff_nh4                   1760                  266        685   
INDUSTRI_S.P.M._tonn     387827                11165       7393   
INDUSTRI_TOC_tonn           272                    0        191   
INDUSTRI_As_tonn              1                    0          0   
INDUSTRI_Pb_tonn              3                    1          2   
INDUSTRI_Cd_tonn              0                    0          0   
INDUSTRI_Cu_tonn              8                    0          1   
INDUSTRI_Zn_tonn             34                    2         14   
INDUSTRI_Ni_tonn             12                    2          6   
INDUSTRI_Cr_tonn              0                    0          0   
INDUSTRI_Hg_tonn              0                    0          0   
RENSEANLEGG_S.P.M._tonn   10780                  675       4294   
RENSEANLEGG_As_tonn           0                    0          0   
RENSEANLEGG_Pb_tonn           0                    0          0   
RENSEANLEGG_Cd_tonn           0                    0          0   
RENSEANLEGG_Cu_tonn           5                    0          2   
RENSEANLEGG_Zn_tonn          15                    0          3   
RENSEANLEGG_Ni_tonn           3                    0          0   
RENSEANLEGG_Cr_tonn           0                    0          0   
RENSEANLEGG_Hg_tonn           0                    0          0   

ospar_region             NORWEGIAN SEA2  SKAGERAK  
flow                             127391     10219  
sew_n                              2498      4530  
sew_p                               343       108  
ind_n                               770      1143  
ind_p                               100        72  
fish_n                            15402        31  
fish_p                             3255         7  
diff_n                            11840      2875  
diff_p                              274        89  
sew_po4                             206        65  
sew_no3                             125       226  
sew_nh4                            1873      3397  
ind_po4                              60        43  
ind_no3                              38        57  
ind_nh4                             577       857  
fish_po4                           2246         5  
fish_no3                           1694         3  
fish_nh4                          12321        25  
diff_po4                             68        22  
diff_no3                           7400      1797  
diff_nh4                            651       158  
INDUSTRI_S.P.M._tonn             367515      1754  
INDUSTRI_TOC_tonn                     2        79  
INDU

## 7. Fish farm copper

Finally, we need to add in the Cu totals from fish farms. The method is similar to that used above, but simpler because we're only interested in one parameter.

In [19]:
engine = rid.connect_to_nivabase(user=user, pw=pw)

# The sql below uses a horrible (and slow!) hack to get around Oracle's
# 1000 item limit on IN clauses. See here for details:
# https://stackoverflow.com/a/9084247/505698
nd_list_hack = [(1, i) for i in nd_list]

sql = (
    "SELECT SUBSTR(regine, 1, 3) AS vassdrag, "
    "  SUM(value) AS value FROM ( "
    "    SELECT b.regine, "
    "           c.name, "
    "           (a.value*d.factor) AS value "
    "    FROM resa2.rid_kilder_aqkult_values a, "
    "    resa2.rid_kilder_aquakultur b, "
    "    resa2.rid_punktkilder_outpar_def c, "
    "    resa2.rid_punktkilder_inp_outp d "
    "    WHERE a.anlegg_nr = b.nr "
    "    AND (1, b.regine) NOT IN %s "
    "    AND a.inp_par_id = d.in_pid "
    "    AND c.out_pid = d.out_pid "
    "    AND name = 'Cu' "
    "    AND ar = %s) "
    "GROUP BY SUBSTR(regine, 1, 3)" % (tuple(nd_list_hack), year)
)

aq_df = pd.read_sql(sql, engine)

if len(aq_df) > 0:
    # Get vassdrag
    aq_df["vass"] = aq_df["vassdrag"].apply(f)
    aq_df = aq_df.query("vass != -999")

    # Calc OSPAR region and group
    aq_df["osp_reg"] = aq_df["vass"].apply(f2)
    aq_df.fillna(0, inplace=True)
    aq_df = aq_df.groupby("osp_reg").sum()
    del aq_df["vass"]

    # Total for Norway
    aq_df.loc["NORWAY"] = aq_df.sum(axis=0)

    # Rename
    aq_df.columns = [
        "AQUAKULTUR_Cu_tonn",
    ]

    # Join model results
    df = df.join(aq_df)

    df.round(0).astype(int).T

Connection successful.


In [20]:
# Write output
out_csv = f"../../../Results/Unmon_loads/teotil2_ospar_unmonitored_loads_{year}.csv"
df.to_csv(out_csv)

This data can then be used to create Table 3 in the report - see [this notebook](https://nbviewer.jupyter.org/github/JamesSample/rid/blob/master/notebooks/summary_table_2017.ipynb) for details.